firefox-ios/Storage/Rust/RustRemoteTabs.swift

```diff
@@ -175,140 +177,162 @@ public class RustRemoteTabs {
+    func getSentCommands(unsentCommandUrls: [String], commands: [PendingCommand]) -> [PendingCommand] {
+        var sentCommands = [PendingCommand]()
+        if unsentCommandUrls.isEmpty {
+            // All of the commands we attempted to send were successfully sent so we do not need to filter
+            // against `unsentCommandUrls`.
+            sentCommands = commands
+        } else {
+            // Filtering the commands we retrieved against the `unsentCommandUrls` we received after
+            // attempting to send close tab commands. This will leave the commands associated with the
+            // `unsentCommandUrls` in the command queue for a future potentially successful close tab
+            // command execution.
+            sentCommands = commands.filter({
+                switch $0.command {
+                case .closeTab(let commandUrl):
+                    return !unsentCommandUrls.contains(commandUrl)
+                }
+            })
+        }
+        return sentCommands
+    }
+
+    private func getUnsentCommandsByDeviceId(deviceId: String, completion: @escaping ([PendingCommand]) -> Void) {
+        commandQueue.async { [unowned self] in
+            guard let tabsCommandQueue = self.tabsCommandQueue else {
+                let err = TabsApiError.UnexpectedTabsError(reason: "Command queue is not initialized") as MaybeErrorType
+                self.logger.log(err.description,
                                 level: .warning,
                                 category: .tabs)
+                completion([PendingCommand]())
+                return
+            }
+
+            tabsCommandQueue.getUnsentCommands { result in
+                switch result {
+                case .success(let commands):
+                    let filteredCommands = commands.filter { $0.deviceId == deviceId }
+                    completion(filteredCommands)
+                case .failure(let error):
+                    self.logger.log("Failed to get unsent commands: \(String(describing: error))",
+                                    level: .warning,
+                                    category: .tabs)
+                    completion([PendingCommand]())
+                }
             }
-        case .failure(let error):
-            self.logger.log("Failed to update command queue: \(String(describing: error))",
-                            level: .warning,
-                            category: .tabs)
         }
     }
 }
 
 internal class RemoteTabsCommandQueue {
-    var commandStore: RemoteCommandStore?
+    var commandStore: RemoteCommandStore
 
-    init() {}
-
-    func openCommandStore(tabsStore: TabsStore) {
+    init(tabsStore: TabsStore) {
         self.commandStore = tabsStore.newRemoteCommandStore()
     }
 
     func addRemoteCommand(deviceId: String,
                           command: RemoteCommand,
                           completion: @escaping (Result<Bool, Error>) -> Void) {
```

Comment:
```suggestion
func addRemoteCommand(deviceId: String, command: RemoteCommand) {
    do {
        let didQueueCommand = try commandStore.addRemoteCommand(deviceId: deviceId, command: command)
        if !didQueueCommand {
            throw TabsApiError.UnexpectedTabsError(reason: "Command already existed")
        }
    } catch {
        self.logger.log("Failed to update command queue: \(String(describing: error))",
                        level: .warning,
                        category: .tabs)
    }
}
```

Could I suggest making this a synchronous API, and pushing the error logging down here? 😅

A completion handler usually indicates asynchronous code, so an API that takes a completion handler, but blocks whatever thread / queue calls it, can be a bit of a surprise.

Moreover, since the only thing we do with the errors in the `Remote Command APIs` methods above is log them, I wonder if we can have the `RemoteTabsCommandQueue` methods look something like:

```swift
func addRemoteCommand(deviceId: String, command: RemoteCommand) {
    do {
        let didQueueCommand = try commandStore.addRemoteCommand(deviceId: deviceId, command: command)
        if !didQueueCommand {
            throw TabsApiError.UnexpectedTabsError(reason: "Command already existed")
        }
    } catch {
        self.logger.log("Failed to update command queue: \(String(describing: error))",
                        level: .warning,
                        category: .tabs)
    }
}
```

...And then simplify the `RustRemoteTabs` methods to something like:

```swift
queue.async { [unowned self] in
    tabsCommandQueue.addRemoteCommand(deviceId: deviceId, command: .closeTab(url: url.absoluteString))
}
```